File: deltafi-api/lib/deltafi/monitor/status/checks/long_running_actions_check.rb
#
#    DeltaFi - Data transformation and enrichment platform
#
#    Copyright 2021-2023 DeltaFi Contributors <deltafi@deltafi.org>
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
#

# frozen_string_literal: true

require 'deltafi/common'
require 'deltafi/monitor/status/check'
require 'time'

module Deltafi
  module Monitor
    module Status
      module Checks
        class LongRunningActionsCheck < Status::Check
          include Deltafi::Logger
          def initialize
            super('Long Running Actions Check')
          end

          def run
            actions = long_running_actions.sort_by { |action, _| action }.to_h
            self.description = "Long Running Actions Check#{" (#{actions.size})" unless actions.empty?}"
            unless actions.empty?
              message_lines << '##### Actions with long running tasks:'
              actions.each do |action, values|
                message_lines << "- #{action}"
                values.sort_by { |value| value[:did] }.each do |value|
                  short_did = value[:did][0..7]
                  did_link = "[#{short_did}](/deltafile/viewer/#{value[:did]})"
                  message_lines << "    - #{did_link} - Running >#{value[:time]} seconds"
                end
              end
            end

            self
          end

          private

          def long_running_actions
            tasks = DF.redis.hgetall(DF::Common::LONG_RUNNING_TASKS_REDIS_KEY)

            tasks.each_with_object(Hash.new { |h, k| h[k] = [] }) do |(k, v), result|
              times_array = JSON.parse(v)
              start_time = Time.parse(times_array[0])
              heartbeat_time = Time.parse(times_array[1])

              next unless Time.now - heartbeat_time < 30

              key_parts = k.split(':')
              action = key_parts[1]
              did = key_parts[2]
              result[action] << { did: did, time: (heartbeat_time - start_time).to_i }
            end
          end
        end
      end
    end
  end
end

File: deltafi-core/src/main/java/org/deltafi/core/monitor/checks/LongRunningActionsCheck.java
/*
 *    DeltaFi - Data transformation and enrichment platform
 *
 *    Copyright 2021-2025 DeltaFi Contributors <deltafi@deltafi.org>
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package org.deltafi.core.monitor.checks;

import org.deltafi.common.types.ActionExecution;
import org.deltafi.core.monitor.MonitorProfile;
import org.deltafi.core.monitor.checks.CheckResult.ResultBuilder;
import org.deltafi.core.services.CoreEventQueue;
import org.deltafi.core.services.DeltaFiPropertiesService;
import org.jetbrains.annotations.NotNull;

import java.util.*;

import static org.deltafi.core.monitor.checks.CheckResult.CODE_YELLOW;

@MonitorProfile
public class LongRunningActionsCheck extends StatusCheck {

    private static final String DESCRIPTION = "Long Running Actions Check";
    private final CoreEventQueue coreEventQueue;
    private final DeltaFiPropertiesService propertiesService;

    public LongRunningActionsCheck(CoreEventQueue coreEventQueue, DeltaFiPropertiesService propertiesService) {
        super(DESCRIPTION);
        this.coreEventQueue = coreEventQueue;
        this.propertiesService = propertiesService;
    }

    @Override
    public CheckResult check() {
        long warningThreshold = propertiesService.getDeltaFiProperties().getActionExecutionWarning() != null ? propertiesService.getDeltaFiProperties().getActionExecutionWarning().getSeconds() : 0L;
        Collection<LongAction> longRunningActions = longRunningActions();
        ResultBuilder resultBuilder = new ResultBuilder();

        if (!longRunningActions.isEmpty()) {
            this.setDescription(DESCRIPTION + " (" + longRunningActions.size() + ")");
            resultBuilder.addHeader("Actions with long running tasks:");
            for (LongAction action : longRunningActions) {
                resultBuilder.addLine("- " + action.name());
                for (Map.Entry<LongActionDetails, Long> didToTime : action.didTimes().entrySet()) {
                    if (warningThreshold != 0L&& didToTime.getValue() >= warningThreshold) {
                        resultBuilder.code(CODE_YELLOW);
                    }
                    String didString = didToTime.getKey().did.toString();
                    String shortDid = didString.substring(0, 7);
                    String didLink = "[" + shortDid + "](/deltafile/viewer/" + didString + ")";
                    resultBuilder.addLine("    - " + didToTime.getKey().appName + " thread " + didToTime.getKey().threadNum + " " + didLink + " - Running >" + didToTime.getValue() + " seconds");
                }
            }
        } else {
            this.setDescription(DESCRIPTION);
        }

        return result(resultBuilder);
    }

    private Collection<LongAction> longRunningActions() {
        Map<String, LongAction> result = new HashMap<>();
        List<ActionExecution> longRunningActions = coreEventQueue.getLongRunningTasks();

        for (ActionExecution actionExecution : longRunningActions) {
            long seconds = actionExecution.heartbeatTime().toEpochSecond() - actionExecution.startTime().toEpochSecond();
            LongAction task = result.computeIfAbsent(actionExecution.action(), LongAction::new);
            task.didTimes().put(new LongActionDetails(actionExecution.did(), actionExecution.threadNum(), actionExecution.appName()), seconds);
        }

        return result.values();
    }

    private record LongAction(String name, Map<LongActionDetails, Long> didTimes) {
        public LongAction(String name) {
            this(name, new TreeMap<>());
        }
    }

    private record LongActionDetails(UUID did, int threadNum, String appName) implements Comparable<LongActionDetails> {
        @Override
        public int compareTo(@NotNull LongActionDetails o) {
            return did.compareTo(o.did) != 0 ? did.compareTo(o.did)
                    : threadNum != o.threadNum ? Integer.compare(threadNum, o.threadNum)
                    : appName.compareTo(o.appName);
        }
    }
}
